### DISCUSSION TITLE
TypeError: 'NoneType' object is not iterable (while registering user)

### DISCUSSION THREAD

**Discussion by xettrisomeman** [fastapi/fastapi#6400] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#routers/user.py
from fastapi import APIRouter, status, HTTPException        
from models.usermodel.user import UserTortoise      

router = APIRouter(
prefix="/users",
tags = ["users"])

@router.post("/", response_model=UserModel)
async def user_create(user: UserCreateModel):
    hashed_password = get_password_hash(user.password)
    try:
        user_tortoise  =await UserTortoise.create(**user.dict(), hashed_password=hashed_password)
    except IntegrityError:
        raise HTTPException(status_code = status.HTTP_400_BAD_REQUEST, detail="Email already exists")
    return UserModel.from_orm(user_tortoise)



#main.py
from routers import user

app = FastAPI()
app.include_router(user.router)


# models/usermodel/user.py
from tortoise.models import Model
from tortoise import fields


class UserTortoise(Model):
    id = fields.IntField(pk=True, generated=True)
    email = fields.CharField(index=True, unique=True,
                             null=False, max_length=255)
    hashed_password = fields.CharField(null=False, max_length=255)

    class Meta:
        table = "users"


# schema/userschema.py

from pydantic import BaseModel
from pydantic.networks import EmailStr


class UserBaseModel(BaseModel):
    email: EmailStr

    class Config:
        orm_mode = True


class UserCreateModel(UserBaseModel):
    password: str


class UserModel(UserBaseModel):
    id: int


class UserDBModel(UserModel):
    hashed_password: str



# database.py

TORTOISE_ORM = {
    "connections": {"default": "sqlite://tortoise.db"},
    "apps": {
        "models": {
            "models": ["models.usermodel.user"],
            "default_connection": "default",
        },
    },
}

register_tortoise(
    app,
    config=TORTOISE_ORM,
    generate_schemas=True,
    add_exception_handlers=True,
)
```

### Description

After doing :

curl -H "accept: application/json" -H "Content-Type: application/json" -X POST http://localhost:8000/users/ -d '{"email" : "hello@gmail.com", "password": "mello"}' 

### Operating System

Linux

### Operating System Details

Ubuntu , 20.04 

### FastAPI Version

0.78.0

### Python Version

3.8.8

### Additional Context

error:
TypeError: 'NoneType' object is not iterable

**Comment by Cajuteq** [comment]
Hi, this is actually pretty difficult to test as the provided code is missing important variables/classes like UserModel, UserCreateModel, ... 
my guess here is that the router you define using APIRouter from fastapi is shadowing or being shadowed by the tortoise/router appearing in your error; 
I'm not familiar with tortoise but you might take a closer look at your imports maybe an "import as router" was written as a shortcut ?

also maybe see the tortoise doc for tortoise [router](https://tortoise-orm.readthedocs.io/en/latest/router.html#usage) can be of help, although it seems to be quite unrelated to what fastapi define as router

**Comment by xettrisomeman** [comment]
> Hi, this is actually pretty difficult to test as the provided code is missing important variables/classes like UserModel, UserCreateModel, ... 

Sorry about that i have added all the required infos. And i tried removing register and added new url and it worked fine. The only problem here is that using tortoise model , when i call fastapi it seems it calls tortoise router and shadow it as you said and when i created pydantic model without tortoise (to test if post request can be sent or not) and sent post request, (inside the user.py) it worked.

**Comment by xettrisomeman** [comment]
solved the problem by using sqlalchemy :|

**Comment by drummerzzz** [comment]
You need create a pyest fixture in `tests/conftest.py` file to load your models before run tests.
Example:

```python
import pytest
from typing import Generator
from fastapi.testclient import TestClient
from tortoise.contrib.test import finalizer, initializer

from main import app

@pytest.fixture(scope='module', autouse=True)
def client() -> Generator:
    initializer(['src.models'])
    with TestClient(app) as test_client:
        yield test_client
    finalizer()
```

**Comment by daudln** [comment]
Register tortoise in main.py.

There for you'r main.py looks like

```py
#main.py
from routers import user

app = FastAPI()
app.include_router(user.router)

register_tortoise(
    app,
    config=TORTOISE_ORM,
    generate_schemas=True,
    add_exception_handlers=True,
)
```
